### PR TITLE
compile -> implementation|api and force -> version strictly

### PR DESCRIPTION
**Pull request by tylerbenson**
and other warnings

There are a few that I left at `compile` in `dd-trace-core`, `dd-trace-ot` and `histograms` because changing them resulted in a broken maven pom file. This should be revisited at some point.

I made some of the dependencies as `api` when they should probably be `implementation` because it simplifies transitive dependencies.

See also: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/643

### CODE REVIEW

**Review by bantonsson — APPROVED** [review]

**Review by lpriima — APPROVED** [review]

File: dd-java-agent/agent-bootstrap/agent-bootstrap.gradle [hunk 1]
```diff
@@ -16,14 +16,14 @@ minimumBranchCoverage = 0.0
...
 
 dependencies {
-  compile project(':dd-java-agent:agent-logging')
-  compile project(':dd-trace-api')
-  compile project(':internal-api')
-  implementation project(':internal-api:internal-api-8')
-  compile deps.slf4j
+  api project(':dd-trace-api')
+  api project(':internal-api')
+  api project(':internal-api:internal-api-8')
```
Comment (cataphract) [inline dd-java-agent/agent-bootstrap/agent-bootstrap.gradle:21]: Any reason this changes to `api`?

File: dd-java-agent/instrumentation/instrumentation.gradle [hunk 2]
```diff
@@ -40,21 +40,21 @@ subprojects { Project subProj ->
       transformation {
         tasks = ['compileJava', 'compileScala', 'compileKotlin']
         plugin = 'datadog.trace.agent.tooling.bytebuddy.AgentGradlePlugin'
-        classPath = project(':dd-java-agent:agent-tooling').configurations.instrumentationMuzzle + subProj.configurations.compile + subProj.sourceSets.main.output
+        classPath = project(':dd-java-agent:agent-tooling').configurations.instrumentationMuzzle + subProj.configurations.compileClasspath + subProj.sourceSets.main.output
```
Comment (cataphract) [inline dd-java-agent/instrumentation/instrumentation.gradle:43]: From what I read above, `instrumentationMuzzle` becomes effectively the `implementation` configuration and should no longer include the compiled classes of the `:dd-java-agent:agent-tooling`. So I'm not sure how this still works

Comment (tylerbenson) [reply dd-java-agent/instrumentation/instrumentation.gradle:43]: hmm... so what do you think it should be instead?

Comment (tylerbenson) [reply dd-java-agent/instrumentation/instrumentation.gradle:43]: I think this is working the way it should even if it's not ideal, so I'm going to merge this and we can revisit if necessary.

**Pull request merged**